src/main/java/io/supertokens/session/accessToken/AccessTokenSigningKey.java

```diff
@@ -69,142 +73,206 @@ public static AccessTokenSigningKey getInstance(Main main) {
+                } else {
+                    for (KeyValueInfo key: keysFromStorage) {
+                        if (config.getAccessTokenSigningKeyDynamic()) {
+                            if (keyCurrentAfter <= key.createdAtTime) {
+                                generateNewKey = false;
+                            }
+                            if (key.createdAtTime < keysValidAfter) {
+                                clearKeys = true;
+                            } else {
+                                validKeysFromSQL.add(new KeyInfo(key.value, key.createdAtTime, validityDuration));
+                            }
+                        }
                     }
                 }
-
-                if (key == null || generateNewKey) {
+                
+                if (generateNewKey) {
                     String signingKey;
                     try {
                         Utils.PubPriKey rsaKeys = Utils.generateNewPubPriKey();
                         signingKey = rsaKeys.toString();
                     } catch (NoSuchAlgorithmException e) {
                         throw new StorageTransactionLogicException(e);
                     }
-                    key = new KeyInfo(signingKey, System.currentTimeMillis());
-                    sqlStorage.setAccessTokenSigningKey_Transaction(con,
-                            new KeyValueInfo(key.value, key.createdAtTime));
+                    KeyInfo newKey = new KeyInfo(signingKey, System.currentTimeMillis(), validityDuration);
+                    sqlStorage.addAccessTokenSigningKey_Transaction(con, new KeyValueInfo(newKey.value, newKey.createdAtTime));
+                    validKeysFromSQL.add(newKey);
                 }
 
-                sqlStorage.commitTransaction(con);
-                return key;
+                if (clearKeys) {
+                    sqlStorage.removeAccessTokenSigningKeysBefore_Transaction(con, keysValidAfter);
+                }
 
+                sqlStorage.commitTransaction(con);
+                return validKeysFromSQL;
             });
         } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
             SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
 
             while (true) {
-
-                KeyInfo key = null;
-                KeyValueInfoWithLastUpdated keyFromStorage = noSQLStorage.getAccessTokenSigningKey_Transaction();
-                if (keyFromStorage != null) {
-                    key = new KeyInfo(keyFromStorage.value, keyFromStorage.createdAtTime);
-                }
-
-                boolean generateNewKey = false;
-
-                if (key != null) {
-                    if (config.getAccessTokenSigningKeyDynamic() && System.currentTimeMillis() > key.createdAtTime
-                            + config.getAccessTokenSigningKeyUpdateInterval()) {
-                        generateNewKey = true;
+                validKeys = new ArrayList<KeyInfo>();
+                long lastCreated = -1;
```

Comment:
I don't quite understand why we need `lastCreated`

Reply:
Aren't we checking the transaction thing by reading the first item of the array in mongo and seeing it's createdTime? - this is why I asked if we need `lastCreated` or not.